sql/core/src/test/scala/org/apache/spark/sql/connector/functions/V2FunctionBenchmark.scala

```diff
@@ -0,0 +1,125 @@
+
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.CodegenObjectFactoryMode._
+import org.apache.spark.sql.connector.catalog.{Identifier, InMemoryCatalog}
+import org.apache.spark.sql.connector.catalog.functions.{BoundFunction, ScalarFunction, UnboundFunction}
+import org.apache.spark.sql.execution.benchmark.SqlBasedBenchmark
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, LongType, StructType}
+
+/**
+ * Benchmark to measure DataSourceV2 UDF performance
+ * {{{
+ *   To run this benchmark:
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class>
+ *        --jars <spark core test jar>,<spark catalyst test jar> <sql core test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/V2FunctionBenchmark-results.txt".
+ * }}}
+ * '''NOTE''': to update the result of this benchmark, please use Github benchmark action:
+ *   https://spark.apache.org/developer-tools.html#github-workflow-benchmarks
+ */
+object V2FunctionBenchmark extends SqlBasedBenchmark {
+  val catalogName: String = "benchmark_catalog"
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    val N = 500L * 1000 * 1000
+    Seq(true, false).foreach { codegenEnabled =>
+      Seq(true, false).foreach { resultNullable =>
+        scalarFunctionBenchmark(N, codegenEnabled = codegenEnabled,
+          resultNullable = resultNullable)
+      }
+    }
+  }
+
+  private def scalarFunctionBenchmark(
+      N: Long,
+      codegenEnabled: Boolean,
+      resultNullable: Boolean): Unit = {
+    withSQLConf(s"spark.sql.catalog.$catalogName" -> classOf[InMemoryCatalog].getName) {
+      createFunction("java_long_add_default",
+        new JavaLongAdd(new JavaLongAddDefault(resultNullable)))
+      createFunction("java_long_add_magic", new JavaLongAdd(new JavaLongAddMagic(resultNullable)))
+      createFunction("java_long_add_static_magic",
+        new JavaLongAdd(new JavaLongAddStaticMagic(resultNullable)))
+      createFunction("scala_long_add_default",
+        LongAddUnbound(new LongAddWithProduceResult(resultNullable)))
+      createFunction("scala_long_add_magic", LongAddUnbound(new LongAddWithMagic(resultNullable)))
+
+      val codeGenFactoryMode = if (codegenEnabled) FALLBACK else NO_CODEGEN
+      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> codegenEnabled.toString,
+          SQLConf.CODEGEN_FACTORY_MODE.key -> codeGenFactoryMode.toString) {
+        val name = s"scalar function (long + long) -> long, result_nullable = $resultNullable " +
+            s"codegen = $codegenEnabled"
+        val benchmark = new Benchmark(name, N, output = output)
+        benchmark.addCase(s"native_long_add", numIters = 3) { _ =>
+          spark.range(N).selectExpr("id + id").noop()
```

Comment:
nit: this is a bit unfair that, native add is always not nullable, even though `resultNullable = true`. We can create a special expression for this benchmark
```
case class NativeAdd(left: Expression, right: Expression, nullable: Boolean) extends BinaryArithmetic {
  ...
}
...
spark.range(N).select(Column(NativeAdd($"id".expr, $"id".expr, resultNullable)))
```

Reply:
Good point. I've updated the benchmark and result.